### DISCUSSION TITLE
Help selecting backend

### DISCUSSION THREAD

**Discussion by rfvalentine** [pyusb/pyusb#443] [Q&A] [answered]
Hello,

pyusb looks like a great option and I would like to use it in my host application to talk to an embedded product.  I am a little confused though on the backend and host driver landscape since there are so many options and some would require my clients to install drivers manually.

My understanding is that libusb1.0 is the latest version of libusb0.1 and they are cross platform.  Right now my host application uses pywinusb to talk to my system via a HID interface.  I want to add a bulk interface to increase my data download (through put) speed between the host application and the embedded system.  Right now everything over HID works great and I want to add an interface that uses bulk packets.

I really need to use a backend that allows me to talk to my embedded system using bulk messages AND I need to continue to use pyinstaller to build a releasable self contained package to my clients.  So my two big host software side requirements are that I can build a self contained archive from pyinstaller with everything needed for a client to unzip and run the host applicaiton which will find the embedded device via PID/VID and connect with no additional driver installation.

So my question ultimately is, what route should one go for a USB 2.0 interface to avoid requiring the end user to install a driver when plugging in the embedded device that communicates over HID and over build packets?  I think the route I should go is pyusb with libusb1.0 backend, but I've seen some posts about drivers still being required on windows 10 using a tool called zadig.  

Right now I only need to deploy my host python software on Windows 10 machines.  Any advice is greatly appreciated as I'm trying to make sure I pick something that will work for me with these requirements.

Thank you!

**Comment by jonasmalacofilho** [comment]
Hi,

These days the right choice 99% of the time is to just use libusb-1.0.  There are some scenarios for which libusb-0.1 is better suited, either because of a limitation in libusb-1.0 or a need to use the libusb(0.1)-win32 driver, but those are super rare.

(While AFAIK libusb-1.0 has indeed evolved from _one of the libusb-0.1 implementations,_ it's ancient history and the libraries are almost irrecognizable, except in APIs that really can't be designed any differently).

> [...] with no additional driver installation.

A driver is necessary for every interface.  While sometimes you can get away with (and it's usually even preferable) using OS supplied drivers (e.g. HIDs), that really depends on what you're working with.

If the OS doesn't have a driver for that bulk interface (technically: interface with bulk endpoint), or if it does but the driver isn't compatible with any of our backends (or any other open-source libraries), then you'll have no choice but to install one.  Regarding user experience, you could mitigate the issue by providing an installer or helper script that does it for them.

>  I think the route I should go is pyusb with libusb1.0 backend, but I've seen some posts about drivers still being required on windows 10 using a tool called zadig.

Apart from the basic requirement that every interface you access have a compatible driver, I've also seen some reports of issues on Windows with composite devices.  Maybe that's what you're referring to?

Anyway, so far I've seen one instance of that being discussed with LibUSB developers (there's a PR, I'll update this with a link if I find it).  The rest, including a couple of patches, have not been yet discussed/proposed with/to the LibUSB devs, and IMO they don't describe the supposed problem well enough for me to tell whether they are legitimate issues/necessary patches.

---

Can you show me that device's USB descriptors (the equivalent of `lsusb -v -d<vid>:<pid>` on Linux)?

**Reply by jonasmalacofilho** [reply]
Oh, I forgot: [the specifics of the Windows backend part of libusb](https://github.com/libusb/libusb/wiki/Windows) might also be useful.

**Reply by mcuee** [reply, marked as answer]
@jonasmalacofilho 
> These days the right choice 99% of the time is to just use libusb-1.0. There are some scenarios for which libusb-0.1 is better suited, either because of a limitation in libusb-1.0 or a need to use the libusb(0.1)-win32 driver, but those are super rare.

Actually using libsubs-win32 driver may no longer be the reason to use libusb-win32 (libusb-0,1 API) any more with the release of libusb-1.0.25/1.0.26 which fixed quite some issues presented in libusb-1.0.23/1.0.24. Based on my recent testing, libusb-1.0+libusb0.sys combination actually work pretty okay (not testing isochronous transfer though -- libusb-1.0 has general weakness for isoc transfer under Windows, even for WinUSB driver, and libusb0.sys has some limitations for isoc transfer support). So even if some legacy device which may work better with libusb0.sys (and not WinUSB), we can still use libusb-1.0 API with those devices.

There are still usage of libusb-0.1 API though.
1) The main reason is that people may not want to change working codes. So they may want to use libusb-0.1 API, either using libusb-compat-0.1 on top of libusb-1.0, or they may want to use libusb-0.1/libusb-win32 (most of the Linux distros now ship libusb-compat-0.1 instead of legacy libusb-0.1).

2) some platform may not have libusb-1.0 support and only libusb-0.1 support (should be very rare now).

3) There are limitations of libusb-1.0 on older systems which leads to the use of the legacy libusb-0.1, for example the following issue for Linux.
* https://github.com/libusb/libusb/issues/237

**Reply by jonasmalacofilho** [reply]
> Actually using libsubs-win32 driver may no longer be the reason to use libusb-win32 [...]

Oh, good to know, thanks!

**Comment by rfvalentine** [comment]
So I have a dev board with the MCU running a bulk endpoint example.  It
seems to enumerate fine and I've captured the traffic below:
[image: image.png]

That is the extent of the traffic when I plug it in and the data
transmission is shown under the summary column.  Also from the screenshot
in the previous email it looks like windows device manager does see the
device.   This is as far as I got before I realized I needed to understand
the host side better because I need to pick a backend for pyusb and really
want to try to get it right before I invest a ton of dev time going down a
dead end.  I guess at this point maybe the most straightforward questions
is the following.

At this point is it possible to pick libusb1.0 and connect to the bulk
endpoints for transmitting data?  I really only need to transmit data from
the device to the host.  Aso you correctly mentioned this will eventually
be a composite device (hopefully).  I am doing the build development first
on a small stand alone dev board and trying to get that figured out before
I merge it with my currently working HID device.  This will ultimately
yield a HID+ bullk composite device.

Sorry if I'm all over the place, but the host side of all this is quite
daunting as I'm an embedded guy.

**Reply by jonasmalacofilho** [reply]
> [image: image.png]

Attachment was stripped by GitHub.  Hint: don't use email replies to GitHub issue/PR/discussion comments, they simply don't work well.

> Also from the screenshot in the previous email [...]

Sorry, I can't see any previous email, or any previous screenshot in this thread.

> At this point is it possible to pick libusb1.0 and connect to the bulk endpoints for transmitting data?

Absolutely, and since you also control the device side of the project, I think you'll be able to work around any eventual issues (e.g. driver issues, or these potential problems with some composite devices) by simply tweaking how you set up its descriptors.

But it might also/still be worth getting some feedback from one of libusb's Windows developers.  (Perhaps even getting your company/project to sponsor them for a while?)

> Sorry if I'm all over the place, but the host side of all this is quite daunting as I'm an embedded guy.

No worries.  And I'm also not super helpful here. I'm not very familiar with the Windows kernel support & drivers.

**Reply by rfvalentine** [reply]
Hi jonas,

Sorry about that.  Here are the images that were left out.

This is a screenshot of my device manager.  It seems the device is showing up, but does have a warning icon over the generic bulk device.
![unnamed](https://user-images.githubusercontent.com/109046385/183821247-9c9daf6d-4658-4e1b-bf49-c76045f003a5.png)

This is a screenshot of the USB traffic capture of the device enumerating.  The more I work on this the more it seems some descriptors are missing.
![unnamed](https://user-images.githubusercontent.com/109046385/183821381-c8c19cb5-913e-4582-983c-1499ccaa76b0.png)

In addition to this, does libusb1.0 require winusb?  It sounds like this is ok as it comes with windows and I only have a single "process" talking to the bulk interface so it seems to fit.  I guess other than possibly having an issue with my eventual goal of a composite device consisting of HID+Bulk.  So far from reading about winusb it sounds like this isn't a problem though.

So I guess from the device manager and scope capture is that telling me I could install libusb1.0 and connect to to the device bulk endpoints?  Sorry I am not very well versed in this area so these questions are probably a little vague and poorly focused.

**Reply by jonasmalacofilho** [reply]
> but does have a warning icon over the generic bulk device.

It would be good to check whether that's because there's no driver attached (in which case you'll likely want to figure out how to have WinUSB attach to it automatically), or because your device is doing something that the driver doesn't like.

> The more I work on this the more it seems some descriptors are missing.

That could explain the warning icon, I think.  But it isn't obvious to me from the second screenshot why you say that...

> In addition to this, does libusb1.0 require winusb?

It requires a kernel driver, and WinUSB is one of the alternatives.  The others are libusb-win32 (the driver) and libusbK (also the driver); although one or both of these will also require additional DLLs.  (There's more information about this in the link I posted before/above).

> So I guess from the device manager and scope capture is that telling me I could install libusb1.0 and connect to to the device bulk endpoints?

I think you can/should certainly try it, although the warning icon suggests things might not work just yet.

**Reply by rfvalentine** [reply]
Hey Jonas,

The reason I said that I think descriptors are missing is because I don't seem them during the enumeration process with my protocol analyzer.  You can see in the screenshot there doesn't seem to be an exchange of any interface descriptors.

**Comment by mcuee** [comment]
@rfvalentine 
libusb wiki may help you to understand a bit better how to use libusb under Windows.
https://github.com/libusb/libusb/wiki/Windows#Driver_Installation

**Reply by rfvalentine** [reply]
I have found quite a few "winusb.dll" files in my system folder.  Is it safe to assume that my windows 10 installation has this by default?  After some researching it seems that winusb ships with windows 10 by default.

As you can see there are quite a few on my system.

![image](https://user-images.githubusercontent.com/109046385/184295351-687dd9f2-c3f6-4d64-a839-a202d01ae7ab.png)

**Reply by mcuee** [reply]
You have to install supported device driver for your device in order to use libusb/pyusb. The presence of winusb.dll is not enough. You can use Zadig.

But you may want to check the reason with the warning icon in Device Manager.
1) no driver installed -- just install WinUSB driver using Zadig and you are good to go.
2) Code 10 or other enumeration error -- you have to fix your device FW

**Reply by mcuee** [reply]
One example using Windows free software tools to troubleshoot.
https://github.com/MicrosoftDocs/windows-driver-docs/blob/staging/windows-driver-docs-pr/usbcon/case-study--troubleshooting-an-unknown-usb-device-by-using-etw-and-netmon.md

But since you have the USB protocol analyzer, you do not need the above.

**Reply by rfvalentine** [reply]
Ok, so it is sounding like it is not possible to develop a custom bulk endpoint interface that can be plug and play like a HID device can be.  Is that correct?  My whole effort really is just trying to get a USB plug and play interface that allows me to transfer data using bulk endpoints rather than interrupt with a HID device.

**Reply by mcuee** [reply]
You can do that with WCID.
https://github.com/pbatard/libwdi/wiki/WCID-Devices

**Reply by rfvalentine** [reply]
mcuee,

Thank you - this looks awesome! I will study this tomorrow!

**Comment by rfvalentine** [comment]
Hi Jonas,

Thank you for the reply! I will get you more info as you requested
tomorrow.  I did however notice that the windows device manager finds a
"generic bulk device" as shown under "other devices".  This is my embedded
device.  You are also correct that I am working on getting to a composite
device.  I have HID working now and want to add the bulk endpoint interface.

My initial hope was just to be able to move data through this bulk pipe
with no need for additional driver installation so I don't have to do any
scripting or write an installer.  Is this generic bulk device an indication
that windows can in fact load a standard driver that I could leverage with
libusb1.0?

[image: image.png]

**Reply by mcuee** [reply]
As  mentioned before, you can use WCID for the sencond generic bulk interface, so that Windows can automatically use the WinUSB driver (at least for Windows 8 and later including Windows 10 and 11).
https://github.com/pbatard/libwdi/wiki/WCID-Devices

An example HID+WinUSB composite device
https://github.com/dioannidis/usbasp
https://github.com/dioannidis/usbasp/blob/master/firmware/usb_descriptors.h

**Reply by jonasmalacofilho** [reply]
> My initial hope was just to be able to move data through this bulk pipe with no need for additional driver installation so I don't have to do any scripting or write an installer. Is this generic bulk device an indication that windows can in fact load a standard driver that I could leverage with libusb1.0?

As @mcuee pointed out, for that you'll likely need WCID.

By the way, at some point you might also hit some issues with composite devices × libusb × Windows.  On a separate project I maintain, we ended up having to use a patched version of [WinUsbCDC](https://gitlab.com/alelec/winusbcdc), instead of PyUSB, to communicate with the WCID bulk interface.  If I'm honest, I'm still quite fuzzy about the details (I'm not that familiar with the Windows drivers, and this was basically about working around multiple issues on lower level APIs), but you can read most of the discussion in liquidctl/liquidctl#479.

> [image: image.png]

Image stripped again.  By the way, email replies to GitHub simply don't work well enough IMO.

**Reply by mcuee** [reply]
@jonasmalacofilho 
I believe the following should help.
* https://github.com/libusb/libusb/pull/965
Unfortunately libusb is lacking Windows developers and reviewer now to move forward with the change.

**Reply by jonasmalacofilho** [reply]
I'm following that PR already, but thanks for mentioning it anyway.

However, it has also been reported to me that that specific patch might not alone be sufficient for some devices... in line with some comments on that PR from user TpImp.

But, again, I'm super lost here, without knowing much about the Windows side of things, and also without a device that exhibits these issues.
